\section{Conclusion}
% \gbwriting{Nanopore sequencing generates noisy electrical signals that need to be converted into a standard string of DNA nucleotide bases using a computational step called basecalling. The performance of basecalling has critical implications for all later steps in genome analysis. Many researchers adopt complex deep learning-based models from the speech recognition domain to perform basecalling without considering the compute demands of such models, which leads to slow, inefficient, and memory-hungry basecallers. Therefore, there is a need to reduce the computation and memory cost of basecalling while maintaining accuracy. However, developing a very fast basecaller that can provide high accuracy requires a deep understanding of genome sequencing, machine learning, and hardware design. Our goal is to develop a comprehensive framework for creating deep learning-based basecallers that provide high efficiency and performance. We introduce \framework, a framework to develop hardware-optimized basecallers. \framework consists of two novel machine-learning techniques that are specifically designed for basecalling: (a) \nas: an automatic architecture search framework that jointly searches for computation blocks in basecaller and best bit-with precision for each neural network layer, and (b) \strim: a dynamic skip removal module to remove hardware-unfriendly skip connections to greatly reduce resource and storage requirements without any loss in basecalling accuracy. We demonstrate the capabilities of \nas and \strim by designing \mech, the first hardware-optimized basecaller that provides both accuracy and inference efficiency. Our extensive evaluation of real genomic organisms shows that \mech provides the ability to basecall quickly, accurately, and efficiently enough to scale the analysis, leading to  $\sim$6.88$\times$ reductions in model size with 2.94$\times$ fewer neural network model parameters.    We hope that our open-source implementations of \nas and \strim inspire future work and ideas in genomics and general omics research and development.}

Nanopore sequencing generates noisy electrical signals that require conversion into a standard DNA nucleotide base string through a computational process known as basecalling. Efficient basecalling is crucial for subsequent genome analysis steps. Current basecalling approaches often neglect computational efficiency, resulting in slow, inefficient, and resource-intensive basecallers. To address this, we present \framework, a framework designed for creating hardware-optimized basecallers. \framework introduces two novel machine-learning techniques: \nas, an automatic architecture search for computation blocks and optimal bit-width precision, and \strim, a dynamic skip connection removal module that significantly reduces resource and storage requirements without sacrificing basecalling accuracy. We demonstrate the capabilities of \nas and \strim by designing
\mech, the first hardware-optimized basecaller, demonstrates fast, accurate, and efficient basecalling, achieving $\sim$6.88$\times$ reductions in model size with 2.94$\times$  fewer neural network parameters compared to an expert designed basecaller. We believe our open-source implementations of \framework will inspire advancements in genomics and omics research and development.