% \documentclass[a4,center,fleqn]{NAR}
% \documentclass[acmsmall]{acmart}[12pt]
\documentclass[camera,hyphens]{jpaper}
\input{packages}
\input{commands}

% Enter dates of publication
% \copyrightyear{YYYY}
% \pubdate{DD MM YYYY}
% \pubyear{YYYY}
% \jvolume{xx}
% \jissue{xx}

% %\articlesubtype{This is the article type (optional)}
% \newcites{supp}{Supplementary References}
% \setlength{\skip\footins}{1.2pc plus 10pt minus 2pt}
% \setlength{\footnotesep}{2pc}
% \includeonly{} %%%%% Uncomment for removing the supplement table
\begin{document}
\sloppy
\title{\ltitle}



\maketitle
\def\thefootnote{*}\footnotetext{\fgb{Correspondence: firtinac@ethz.ch; omutlu@ethz.ch}}\def\thefootnote{\arabic{footnote}}
\input{0_abstract}
\section*{Keywords}
genomics sequencing, basecalling, hardware acceleration, machine learning, deep neural network
\input{1_introduction}
\input{2_results}
% \input{sections/motivation}
\input{3_discussion}
\input{4_conclusion}
\input{5_evaluation}

% \input{sections/5_conclusion}

\section{Availability of data and materials}
\label{sec:data_avail}
The read set and reference set used in this study \fgb{are part of work carried out by Wick et al.~\cite{wick2019performance},} which can be downloaded from \url{https://bridges.monash.edu/articles/dataset/Raw_fast5s/7676174} and \url{https://bridges.monash.edu/articles/dataset/Reference_genomes/7676135}, respectively. 
\Copy{CC1/1}{\hgb{For the human genome~\cite{rhie2023complete}, we download reads from \mbox{\url{https://labs.epi2me.io/gm24385_2020.11/}}, while the reference genome is available at \mbox{\url{https://github.com/marbl/HG002}}.}} All trained models and generated reads can be downloaded from \url{https://zenodo.org/record/10198815}.
We ensure unbiased, fair, and consistent evaluation by retraining all the basecallers using the official ONT dataset~\mbox{\cite{bonito}}.

% \section{Code Availability}
\hgb{Source code with the instructions for reproducing the results is publicly available at: 
 \fgb{GitHub~\cite{rubiconCode} and Zenodo~\cite{zenadoCode}}}.
%Source code with the instructions for reproducing the results is shared and can be downloaded using \url{https://figshare.com/s/98b1396a85155c3f8b52} during the review process and will be published on GitHub upon acceptance to help researchers develop future hardware-optimized basecallers. We provide installation scripts for both NVIDIA and AMD GPUs.
Scripts used to perform basecalling accuracy analysis are available at: \url{https://github.com/rrwick/Basecalling-comparison}.
 %We will make it publicly available upon publication.
%https://github.com/CMU-SAFARI/RUBICON.


\section{Acknowledgments}
\gsssss{We thank the SAFARI Research Group members for their valuable feedback and the stimulating intellectual \gssssss{and} scholarly environment they provide. SAFARI Research Group acknowledges the generous gifts of their industrial partners, including Google, Huawei, Intel, Microsoft, VMware, and AMD. This research was partially supported by the Semiconductor Research Corporation. SAFARI Research Group acknowledges support from the European Unionâ€™s Horizon programme for research and innovation under grant agreement No. 101047160, project BioPIM.}
 Special thanks to  Alessandro Pappalardo for his support with quantization-aware training. We appreciate valuable discussions with Giovanni Mariani.  Thanks to AMD for providing access to the HPC fund cluster~\cite{hpcfund}. %We thank the SAFARI Research Group members for valuable feedback and the stimulating intellectual environment they provide. 

\section{Author contributions}
 G.S., M.A., K.D., and A.K.  conceived \framework. G.S. designed and implemented \framework. %G.S., C.F., and M.C. collected data.  
 G.S., C.F., and M.C. collected data and performed the evaluations.  K.D., H.C., and O.M. supervised the work. %K.D. and O.M. acquired funding and supervised the project.  
 G.S., M.A.,  K.D., and C.F. wrote the manuscript. %G.S. produced the figures. 
All authors reviewed and edited the manuscript. All authors analyzed the results. All authors read and approved the final manuscript.  

\fgb{
\section{Ethics approval and consent to participate}
Not applicable - ethical approval was not needed for the study, as publicly available datasets were used. No private, confidential, or sensitive information pertaining to individuals was utilized. Furthermore, our research did not involve any animal experiments.
}
\section{Competing interests} 
\fgb{Gagandeep Singh, Kristof Denolf, and Alireza Khodamoradi are affiliated with AMD.}
The remaining authors declare no competing interests.

\fgb{
\section{Supplementary information}
Additional 1: Supplementary notes S1-S6, Figures S1-S9, and Tables S1-S3.
}
% \fgb{
% \section{Corresponding author}
% Correspondence to Gagandeep Singh.}

% \section{Funding}
% O.M. and SAFARI Research Group are supported by funding from Intel, VMware, Semiconductor Research Corporation, the National Institutes of Health (NIH), and the ETH Future Computing Laboratory.
% % }


\newpage
\balance
% \bibliographystyle{IEEEtranS}
\bibliographystyle{vancouver}

{\small \bibliography{ref}}
\input{sections/supp}
% \input{revision}
\end{document}
